# TMDb Movie Data Analysis

## Introduction

This data set contains information about 10,000 movies collected from The Movie Database (TMDb), including user ratings and revenue.

- The ‘cast’ and ‘genres’, contain multiple values separated by pipe (|) characters.
- There are some odd characters in the ‘cast’ column. Don’t worry about cleaning them. You can leave them as is.
- The final two columns ending with “_adj” show the budget and revenue of the associated movie in terms of 2010 dollars, accounting for inflation over time.


__Questions to research in the analysis__
- Which genres are most popular from year to year?
- What kinds of properties are associated with movies that have high revenues?
- Directors with most movies
- Most Popular genres
- Most earning / popular movie in a year


> Max budget, max revenue, max profit, oldest movie

In [6]:
# Import libraries

import pandas as pd  
import numpy as np 
import matplotlib.pyplot as plt  
%matplotlib inline

In [7]:
df = pd.read_csv('tmdb-movies.csv')

In [8]:
# Visual Assessment 
df.head(2)

id    imdb_id  popularity     budget     revenue      original_title  \
0  135397  tt0369610   32.985763  150000000  1513528810      Jurassic World   
1   76341  tt1392190   28.419936  150000000   378436354  Mad Max: Fury Road   

                                                cast  \
0  Chris Pratt|Bryce Dallas Howard|Irrfan Khan|Vi...   
1  Tom Hardy|Charlize Theron|Hugh Keays-Byrne|Nic...   

                        homepage         director             tagline  ...  \
0  http://www.jurassicworld.com/  Colin Trevorrow   The park is open.  ...   
1    http://www.madmaxmovie.com/    George Miller  What a Lovely Day.  ...   

                                            overview runtime  \
0  Twenty-two years after the events of Jurassic ...     124   
1  An apocalyptic story set in the furthest reach...     120   

                                      genres  \
0  Action|Adventure|Science Fiction|Thriller   
1  Action|Adventure|Science Fiction|Thriller   

                                production_companies release_date vote_count  \
0  Universal Studios|Amblin Entertainment|Legenda...       6/9/15       5562   
1  Village Roadshow Pictures|Kennedy Miller Produ...      5/13/15       6185   

   vote_average  release_year    budget_adj   revenue_adj  
0           6.5          2015  1.379999e+08  1.392446e+09  
1           7.1          2015  1.379999e+08  3.481613e+08  

[2 rows x 21 columns]

In [9]:
# Column information & Null Values
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10866 entries, 0 to 10865
Data columns (total 21 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    10866 non-null  int64  
 1   imdb_id               10856 non-null  object 
 2   popularity            10866 non-null  float64
 3   budget                10866 non-null  int64  
 4   revenue               10866 non-null  int64  
 5   original_title        10866 non-null  object 
 6   cast                  10790 non-null  object 
 7   homepage              2936 non-null   object 
 8   director              10822 non-null  object 
 9   tagline               8042 non-null   object 
 10  keywords              9373 non-null   object 
 11  overview              10862 non-null  object 
 12  runtime               10866 non-null  int64  
 13  genres                10843 non-null  object 
 14  production_companies  9836 non-null   object 
 15  release_date       

- Few Null values exist in `cast`, `director`, `overview` & `genres`, so these can be dropped
- Many Null values exist in `homepage`, `tagline`, `keywords` & `production_companies` columns. Further we can drop `keywords` & `production_companies` since this data is not necesary in the analysis.

Descriptive Statistics

In [10]:
df.describe()

id    popularity        budget       revenue       runtime  \
count   10866.000000  10866.000000  1.086600e+04  1.086600e+04  10866.000000   
mean    66064.177434      0.646441  1.462570e+07  3.982332e+07    102.070863   
std     92130.136561      1.000185  3.091321e+07  1.170035e+08     31.381405   
min         5.000000      0.000065  0.000000e+00  0.000000e+00      0.000000   
25%     10596.250000      0.207583  0.000000e+00  0.000000e+00     90.000000   
50%     20669.000000      0.383856  0.000000e+00  0.000000e+00     99.000000   
75%     75610.000000      0.713817  1.500000e+07  2.400000e+07    111.000000   
max    417859.000000     32.985763  4.250000e+08  2.781506e+09    900.000000   

         vote_count  vote_average  release_year    budget_adj   revenue_adj  
count  10866.000000  10866.000000  10866.000000  1.086600e+04  1.086600e+04  
mean     217.389748      5.974922   2001.322658  1.755104e+07  5.136436e+07  
std      575.619058      0.935142     12.812941  3.430616e+07  1.446325e+08  
min       10.000000      1.500000   1960.000000  0.000000e+00  0.000000e+00  
25%       17.000000      5.400000   1995.000000  0.000000e+00  0.000000e+00  
50%       38.000000      6.000000   2006.000000  0.000000e+00  0.000000e+00  
75%      145.750000      6.600000   2011.000000  2.085325e+07  3.369710e+07  
max     9767.000000      9.200000   2015.000000  4.250000e+08  2.827124e+09

## Data Wrangling

### Gathering Data: 
The data is imported from CSV file into the notebook within a pandas dataframe

### Assessing Data:

This [link](https://www.themoviedb.org/talk/5141d424760ee34da71431b0) states that `popularity` score is measured by counting unique views on the website with number of ratings, favourites & watched list additions. Also, it has no upper bond. 

In [11]:
# Filter budget data 

df_zero_budget = df.query('budget == 0')
df_zero_budget.head()

id    imdb_id  popularity  budget   revenue  \
30  280996  tt3168230    3.927333       0  29355203   
36  339527  tt1291570    3.358321       0  22354572   
72  284289  tt2911668    2.272044       0     45895   
74  347096  tt3478232    2.165433       0         0   
75  308369  tt2582496    2.141506       0         0   

                    original_title  \
30                      Mr. Holmes   
36                          Solace   
72                Beyond the Reach   
74          Mythica: The Darkspore   
75  Me and Earl and the Dying Girl   

                                                 cast  \
30  Ian McKellen|Milo Parker|Laura Linney|Hattie M...   
36  Abbie Cornish|Jeffrey Dean Morgan|Colin Farrel...   
72  Michael Douglas|Jeremy Irvine|Hanna Mangan Law...   
74  Melanie Stone|Kevin Sorbo|Adam Johnson|Jake St...   
75  Thomas Mann|RJ Cyler|Olivia Cooke|Connie Britt...   

                                             homepage  \
30                       http://www.mrholmesfilm.com/   
36                                                NaN   
72                                                NaN   
74          http://www.mythicamovie.com/#!blank/wufvh   
75  http://www.foxsearchlight.com/meandearlandthed...   

                   director  \
30              Bill Condon   
36            Afonso Poyart   
72  Jean-Baptiste Léonetti   
74            Anne K. Black   
75      Alfonso Gomez-Rejon   

                                              tagline  ...  \
30                            The man behind the myth  ...   
36  A serial killer who can see your future, a psy...  ...   
72                                                NaN  ...   
74                                                NaN  ...   
75           A Little Friendship Never Killed Anyone.  ...   

                                             overview runtime  \
30  The story is set in 1947, following a long-ret...     103   
36  A psychic doctor, John Clancy, works with an F...     101   
72  A high-rolling corporate shark and his impover...      95   
74  When Teela’s sister is murdered and a powerf...     108   
75  Greg is coasting through senior year of high s...     105   

                      genres  \
30             Mystery|Drama   
36       Crime|Drama|Mystery   
72                  Thriller   
74  Action|Adventure|Fantasy   
75              Comedy|Drama   

                                 production_companies release_date vote_count  \
30  BBC Films|See-Saw Films|FilmNation Entertainme...      6/19/15        425   
36  Eden Rock Media|FilmNation Entertainment|Flynn...       9/3/15        474   
72                                      Furthur Films      4/17/15         81   
74                           Arrowstorm Entertainment      6/24/15         27   
75                                  Indian Paintbrush      6/12/15        569   

    vote_average  release_year  budget_adj   revenue_adj  
30           6.4          2015         0.0  2.700677e+07  
36           6.2          2015         0.0  2.056620e+07  
72           5.5          2015         0.0  4.222338e+04  
74           5.1          2015         0.0  0.000000e+00  
75           7.7          2015         0.0  0.000000e+00  

[5 rows x 21 columns]

- Upon searching the internet, found that **Beyond the Reach** movie had a budget of $ 396k

In [12]:
# Filter data
df_zero_revenue = df.query('revenue == 0')
df_zero_revenue.head()

id    imdb_id  popularity    budget  revenue  \
48  265208  tt2231253    2.932340  30000000        0   
67  334074  tt3247714    2.331636  20000000        0   
74  347096  tt3478232    2.165433         0        0   
75  308369  tt2582496    2.141506         0        0   
92  370687  tt3608646    1.876037         0        0   

                    original_title  \
48                       Wild Card   
67                        Survivor   
74          Mythica: The Darkspore   
75  Me and Earl and the Dying Girl   
92        Mythica: The Necromancer   

                                                 cast  \
48  Jason Statham|Michael Angarano|Milo Ventimigli...   
67  Pierce Brosnan|Milla Jovovich|Dylan McDermott|...   
74  Melanie Stone|Kevin Sorbo|Adam Johnson|Jake St...   
75  Thomas Mann|RJ Cyler|Olivia Cooke|Connie Britt...   
92  Melanie Stone|Adam Johnson|Kevin Sorbo|Nicola ...   

                                             homepage             director  \
48                                                NaN           Simon West   
67                          http://survivormovie.com/       James McTeigue   
74          http://www.mythicamovie.com/#!blank/wufvh        Anne K. Black   
75  http://www.foxsearchlight.com/meandearlandthed...  Alfonso Gomez-Rejon   
92          http://www.mythicamovie.com/#!blank/y9ake        A. Todd Smith   

                                        tagline  ...  \
48  Never bet against a man with a killer hand.  ...   
67           His Next Target is Now Hunting Him  ...   
74                                          NaN  ...   
75     A Little Friendship Never Killed Anyone.  ...   
92                                          NaN  ...   

                                             overview runtime  \
48  When a Las Vegas bodyguard with lethal skills ...      92   
67  A Foreign Service Officer in London tries to p...      96   
74  When Teela’s sister is murdered and a powerf...     108   
75  Greg is coasting through senior year of high s...     105   
92  Mallister takes Thane prisoner and forces Mare...       0   

                      genres  \
48      Thriller|Crime|Drama   
67     Crime|Thriller|Action   
74  Action|Adventure|Fantasy   
75              Comedy|Drama   
92  Fantasy|Action|Adventure   

                                 production_companies release_date vote_count  \
48  Current Entertainment|Lionsgate|Sierra / Affin...      1/14/15        481   
67  Nu Image Films|Winkler Films|Millennium Films|...      5/21/15        280   
74                           Arrowstorm Entertainment      6/24/15         27   
75                                  Indian Paintbrush      6/12/15        569   
92  Arrowstorm Entertainment|Camera 40 Productions...     12/19/15         11   

    vote_average  release_year    budget_adj  revenue_adj  
48           5.3          2015  2.759999e+07          0.0  
67           5.4          2015  1.839999e+07          0.0  
74           5.1          2015  0.000000e+00          0.0  
75           7.7          2015  0.000000e+00          0.0  
92           5.4          2015  0.000000e+00          0.0  

[5 rows x 21 columns]

- [Wikipedia](https://en.wikipedia.org/wiki/Wild_Card_(2015_film)) states that **Wild Card** had budget & box office collecion.
- The movie **Mythica: The Darkspore** received a [tax credit](https://en.wikipedia.org/wiki/Mythica_(film_series)) for filming in Governor's office

Hence, the zero  values in `budget` & `revenue` columns are missing values in the data set.

Let's count the total zero values. Based on the total number of missing values we can drop or retain them.

In [13]:
print("Total missing values in budget column:",df_zero_budget.shape[0])
print("Total missing values in revenue column:",df_zero_revenue.shape[0])


Total missing values in budget column: 5696
Total missing values in revenue column: 6016


Almost 50% of the rows in the dataset have missing values in the `budget` & `revenue` columns respectively.
Instead of Dropping, we replace the missing with Null values.

In [14]:
df.query('runtime == 0').count()

id                      31
imdb_id                 31
popularity              31
budget                  31
revenue                 31
original_title          31
cast                    31
homepage                 6
director                29
tagline                  5
keywords                15
overview                29
runtime                 31
genres                  30
production_companies    13
release_date            31
vote_count              31
vote_average            31
release_year            31
budget_adj              31
revenue_adj             31
dtype: int64

#### Issues
- Drop unnecessary columns 
- Drop duplicates
- Drop Null values from `cast`, `director`, `genres`
- Fix: Zero values in `budget` & `revenue` columns.
- Drop zero value records in `runtime`
- Fix `release_date` dtype

- Split values in `genre` & `casts` into separate columns.

### Cleaning Data:

In [15]:
# Define: Drop unnecesary columns

# Code
col = ['imdb_id', 'homepage', 'tagline', 'overview', 'budget_adj', 'revenue_adj']
df.drop(col, axis = 1, inplace = True)

# Test
df.columns

Index(['id', 'popularity', 'budget', 'revenue', 'original_title', 'cast',
       'director', 'keywords', 'runtime', 'genres', 'production_companies',
       'release_date', 'vote_count', 'vote_average', 'release_year'],
      dtype='object')

In [16]:
# Define: Drop duplicated values

# Code
df.drop_duplicates(inplace=True)

# Test 
df.duplicated().any()

False

In [17]:
# Define: Drop Null values from cast, director, genres columns

# Code
cols = ['cast', 'director', 'genres']
df.dropna(subset = cols, how = 'any', inplace = True)

# Test
df.isnull().sum()

id                         0
popularity                 0
budget                     0
revenue                    0
original_title             0
cast                       0
director                   0
keywords                1425
runtime                    0
genres                     0
production_companies     959
release_date               0
vote_count                 0
vote_average               0
release_year               0
dtype: int64

In [18]:
#Define: Zero values in `budget` & `revenue` column

# Code

df['budget'] = df['budget'].replace(0, np.NaN)
df['revenue'] = df['revenue'].replace(0, np.NaN)

# Test
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10731 entries, 0 to 10865
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   id                    10731 non-null  int64  
 1   popularity            10731 non-null  float64
 2   budget                5153 non-null   float64
 3   revenue               4843 non-null   float64
 4   original_title        10731 non-null  object 
 5   cast                  10731 non-null  object 
 6   director              10731 non-null  object 
 7   keywords              9306 non-null   object 
 8   runtime               10731 non-null  int64  
 9   genres                10731 non-null  object 
 10  production_companies  9772 non-null   object 
 11  release_date          10731 non-null  object 
 12  vote_count            10731 non-null  int64  
 13  vote_average          10731 non-null  float64
 14  release_year          10731 non-null  int64  
dtypes: float64(4), int6

The above result shows, 5153 zero values in `budget` are replaced with Null values. Similar action was executed for `revenue` column.

In [19]:
# Define: Drop xero values in 'runtime' column

# Code
df.query('runtime != 0', inplace = True)

# Test
df.query('runtime == 0')

Empty DataFrame
Columns: [id, popularity, budget, revenue, original_title, cast, director, keywords, runtime, genres, production_companies, release_date, vote_count, vote_average, release_year]
Index: []

In [20]:
# Fix 'release_date' dtype
df['release_date'] = pd.to_datetime(df['release_date'])

# Check
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 10703 entries, 0 to 10865
Data columns (total 15 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   id                    10703 non-null  int64         
 1   popularity            10703 non-null  float64       
 2   budget                5150 non-null   float64       
 3   revenue               4843 non-null   float64       
 4   original_title        10703 non-null  object        
 5   cast                  10703 non-null  object        
 6   director              10703 non-null  object        
 7   keywords              9293 non-null   object        
 8   runtime               10703 non-null  int64         
 9   genres                10703 non-null  object        
 10  production_companies  9759 non-null   object        
 11  release_date          10703 non-null  datetime64[ns]
 12  vote_count            10703 non-null  int64         
 13  vote_average    